apps/src/lib/tools/jsinterpreter/CustomMarshalingInterpreter.js

```diff
@@ -263,22 +270,54 @@ export default class CustomMarshalingInterpreter extends Interpreter {
     return super.setProperty(...args);
   }
 
+  /**
+   * Execute one step of the interpreter.
+   * @return {boolean} True if a step was executed, false if no more instructions.
+   */
   step() {
-    const state = this.peekStackFrame();
-    // Program nodes always have end=0 for some reason (acorn related).
-    // The Interpreter.step method assumes that a falsey state.node.end value means
-    // the interpreter is inside polyfill code, because it strips all location information from ast nodes for polyfill code.
-    // This means the interpreter will sometimes step more often than necessary. This is a problem for us when breakpoints
-    // are turned on because the interpreter can step over nodes that we need to check before they get stepped, resulting
-    // in an infinite loop.
+    //
+    // The Interpreter.step method recurses to try to make the step() operation do more when
+    // calling polyfill code. This recursion can cause exceptions with the maximum call stack
+    // size is exceeded.
+    //
     // See this line in the interpreter code which introduced this behavior:
     //   https://github.com/NeilFraser/JS-Interpreter/commit/a4ded3ed1de7960cda9177d1bacb6a2526440d14#diff-966ad2ec9f775b3820dd37b4d36b650aR116
-    // TODO: push a fix upstream that checks state.node.end === undefined so the interpreter
-    // doesn't step unnecessarily for Program nodes.
-    if (state && state.node.type === 'Program') {
-      state.node.end = 1;
+    //
+    // TODO: push a fix upstream that executes the function body in a while loop instead of
+    // recursing by calling step() within step()
+    //
+    // We have cloned the functionality of Interpreter.step() here with the offending portion
+    // disabled through comments
+    //
```

Comment:
Why are we fixing this in CustomMarshalingInterpreter.js instead of in https://github.com/code-dot-org/JS-Interpreter/blob/master/interpreter.js, when polyfills are a concern in that external dependency, not specific to custom marshaling?

Reply:
fair point - I had forgotten that we have a fork of `intepreter.js` - happy to fix there, but I'll probably leave the regression test here. agreed?

Reply:
In that case, I'd like to remove the `CustomMarshalingInterpreter` implementation of `step()` since we don't need to run this code during every `step()` operation:

```
if (state && state.node.type === 'Program') {
      state.node.end = 1;
```

We can do it in the constructor as I've done in this PR. So, my plan is to keep this PR around, but adapt it to use a new version of `intepreter.js` from our fork.

Reply:
:+1: That sounds great!

Reply:
PR updated with new version of the interpreter